pkg/filters/string.go

```diff
@@ -6,55 +6,174 @@ import (
-		}
-		operatorString = operatorAndValues[0:2]
-		valuesString = operatorAndValues[2:]
+	if len(filters) > 0 {
+		filter.Enable()
 	}
 
-	values := strings.Split(valuesString, ",")
+	return filter, nil
+}
 
-	for i := range values {
-		switch operatorString {
-		case "=":
-			filter.Equal = append(filter.Equal, values[i])
-		case "!=":
-			filter.NotEqual = append(filter.NotEqual, values[i])
-		default:
-			return fmt.Errorf("invalid filter operator: %s", operatorString)
+// priority goes by (from most significant):
+// 1. equality, suffixed, prefixed, contains
+// 2. not equals, not suffixed, not prefixed, not contains
+// This is done so if a conflicting "not" filter exists, we ignore it
+func (f *StringFilter) Filter(val string) bool {
+	enabled := f.enabled
+	equals := f.equal[val]
+	notEquals := f.notEqual[val]
+	suffixes := f.suffixes
+	prefixes := f.prefixes
+	contains := f.contains
+	notSuffixes := f.notSuffixes
+	notPrefixes := f.notPrefixes
+	notContains := f.notContains
+	notEqualsSet := len(f.notEqual) > 0 || notSuffixes.Length() > 0 || notPrefixes.Length() > 0 || len(notContains) > 0
+	if !enabled {
+		return true
+	}
+	if equals {
+		return true
+	}
+	if suffixes.Filter(val) {
+		return true
+	}
+	if prefixes.Filter(val) {
+		return true
+	}
+	for contain := range contains {
+		if strings.Contains(val, contain) {
+			return true
+		}
+	}
+	if notEqualsSet {
+		if notSuffixes.Filter(val) {
+			return false
+		}
+		if notPrefixes.Filter(val) {
+			return false
 		}
+		for contain := range notContains {
+			if strings.Contains(val, contain) {
+				return false
+			}
+		}
+		res := !notEquals || equals
+		return res
 	}
+	return false
+}
 
-	filter.Enable()
+func (f *StringFilter) parse(filterReq protocol.Filter) error {
+	arr := filterReq.Value
+	vals := make([]string, len(filterReq.Value))
+	for i := 0; i < len(arr); i++ {
+		val := arr[i]
+		valStr, ok := val.(string)
+		if !ok {
+			return fmt.Errorf("failed to add to filter: invalid value: %s", val)
+		}
+		vals[i] = valStr
+	}
+	for _, val := range vals {
+		err := f.add(val, Operator(filterReq.Operator))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
 
+func (f *StringFilter) add(val string, operator Operator) error {
+	switch operator {
+	case Equal:
+		f.addEqual(val)
+	case NotEqual:
+		f.addNotEqual(val)
+	default:
+		return UnsupportedOperator(operator)
+	}
 	return nil
 }
 
+func (f *StringFilter) addEqual(val string) {
+	prefixSet := val[len(val)-1] == '*'
+	suffixSet := val[0] == '*'
+	if prefixSet && suffixSet && len(val) > 1 {
+		f.contains[val[1:len(val)-1]] = true
+		return
+	}
+	if prefixSet {
+		f.prefixes.Put(val[:len(val)-1])
+		return
+	}
+	if suffixSet && len(val) > 1 {
+		f.suffixes.Put(val[1:])
+		return
+	}
+	if !prefixSet && !suffixSet {
+		f.equal[val] = true
+		return
+	}
+}
+
+func (f *StringFilter) addNotEqual(val string) {
+	prefixSet := val[len(val)-1] == '*'
+	suffixSet := val[0] == '*'
+	if prefixSet {
```

Comment:
We should probably check that either `*` and `**` are not given here as val

Reply:
Did that and added tests